quinn/src/lib.rs

```diff
@@ -805,13 +796,20 @@ impl Drop for ConnectionInner {
                 return;
             }
             pending.get_mut().dropped = true;
-            if pending.get().closing.is_none() {
+            if !pending.get().closing {
                 endpoint
                     .inner
                     .close(Instant::now(), self.handle, 0, (&[][..]).into());
                 if let Some(x) = endpoint.driver.as_ref() {
                     x.notify();
                 }
+                pending
+                    .get_mut()
+                    .fail(ConnectionError::TransportError(quinn::TransportError {
+                        code: quinn::TransportErrorCode::INTERNAL_ERROR,
+                        frame: None,
+                        reason: "driver future was dropped".to_string(),
```

Comment:
I find this error a bit confusing. In my mind the `driver` concept is associated with a particular thing (in our case only the `Driver`), where in this case all references to the `ConnectionInner` have been dropped. I think these are held by `Connection` (which I wouldn't really count as a driver) and `IncomingStreams` (same).

Maybe something like `all references to the connection were dropped`?

Also, what's our take on a `Connection` being alive without `IncomingStreams` (should be okay I think) or an `IncomingStreams` being alive without the `Connection` (maybe more iffy?)?

Reply:
> I find this error a bit confusing.

Whoops, certainly did not intend to copy that error message verbatim. Good catch!

> Also, what's our take on a Connection being alive without IncomingStreams (should be okay I think) or an IncomingStreams being alive without the Connection (maybe more iffy?)?

I think these are both reasonable cases. For the `Connection`-only case, not all application protocols will involve streams in both directions being continually opened, and for `IncomingStreams`-only, an application layer might not have any use for the various accessors once the connection gets going in earnest, so forcing them to keep the `Connection` around would be unnatural.